luigi/contrib/external_program.py

```diff
@@ -0,0 +1,159 @@
+
+            stdout = self._clean_output_file(tmp_stdout)
+            stderr = self._clean_output_file(tmp_stderr)
+
+            if stdout:
+                logger.info('Program stdout:\n{}'.format(stdout))
+            if stderr:
+                logger.info('Program stderr:\n{}'.format(stderr))
+
+            if proc.returncode != 0:
+                raise ExternalProgramRunError(
+                    'Program failed with return code={}:'.format(proc.returncode),
+                    args, env=env, stdout=stdout, stderr=stderr)
+        finally:
+            tmp_stderr.close()
+            tmp_stdout.close()
+
+
+class ExternalProgramRunContext(object):
+    def __init__(self, proc):
+        self.proc = proc
+
+    def __enter__(self):
+        self.__old_signal = signal.getsignal(signal.SIGTERM)
+        signal.signal(signal.SIGTERM, self.kill_job)
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if exc_type is KeyboardInterrupt:
+            self.kill_job()
+        signal.signal(signal.SIGTERM, self.__old_signal)
+
+    def kill_job(self, captured_signal=None, stack_frame=None):
+        self.proc.kill()
+        if captured_signal is not None:
+            # adding 128 gives the exit code corresponding to a signal
+            sys.exit(128 + captured_signal)
+
+
+class ExternalProgramRunError(RuntimeError):
+    def __init__(self, message, args, env=None, stdout=None, stderr=None):
+        super(ExternalProgramRunError, self).__init__(message, args, env, stdout, stderr)
+        self.message = message
+        self.args = args
+        self.env = env
+        self.out = stdout
+        self.err = stderr
+
+    def __str__(self):
+        info = self.message
+        info += '\nCOMMAND: {}'.format(' '.join(self.args))
+        info += '\nSTDOUT: {}'.format(self.out or '[empty]')
+        info += '\nSTDERR: {}'.format(self.err or '[empty]')
+        env_string = None
+        if self.env:
+            env_string = ' '.join(['='.join([k, '\'{}\''.format(v)]) for k, v in self.env.items()])
+        info += '\nENVIRONMENT: {}'.format(env_string or '[empty]')
+        return info
+
+
+class ExternalPythonProgramTask(ExternalProgramTask):
+    """
+    Template task for running an external Python program in a subprocess
+
+    Simple extension of :py:class:`ExternalProgramTask`, adding two
+    :py:class:`luigi.parameter.Parameter` s for setting a virtualenv and for
+    extending the ``PYTHONPATH``.
+    """
+    virtualenv = luigi.Parameter(
+        default=None,
+        description='path to the virtualenv directory to use. It should point to '
+                    'the directory containing the ``bin/activate`` file used for '
+                    'enabling the virtualenv.')
+    extra_pythonpath = luigi.Parameter(
+        default=None,
+        description='extend the search path for modules by prepending this '
+                    'value to the ``PYTHONPATH`` environment variable.')
```

Comment:
You must set `positional=False` for these two parameters. Because this is a typical base task.